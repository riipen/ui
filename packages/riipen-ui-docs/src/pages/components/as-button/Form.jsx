import React, { useState } from "react";

import AsButton from "riipen-ui/components/AsButton";

export default function Form() {
  const [formState, setFormState] = useState({});
  const [message, setMessage] = useState("Click a card to submit the form");

  const style = {
    marginRight: "10px"
  };

  const cardStyle = {
    boxShadow: "1px 1px 1px 1px",
    padding: "10px",
    ...style
  };

  const handleSubmit = event => {
    event.preventDefault();
    setMessage(`You submitted the form and chose: ${formState.selectedEntity}`);
  };

  const handleClick = entity => {
    setFormState({
      selectedEntity: entity.id
    });
  };

  const entities = [
    { id: 1, title: "Entity One", details: "Details One" },
    { id: 2, title: "Entity Two", details: "Details 2" }
  ];

  return (
    <>
      <form onSubmit={handleSubmit}>
        {entities.map(entity => (
          <AsButton
            ariaLabel={`Click to select ${entity.title} and submit the form`}
            key={entity.id}
            onClick={() => handleClick(entity)}
            type="submit"
          >
            <div style={cardStyle}>
              <h3>{entity.title}</h3>
              <p>{entity.details}</p>
            </div>
          </AsButton>
        ))}
      </form>
      <p>{message}</p>
    </>
  );
}
