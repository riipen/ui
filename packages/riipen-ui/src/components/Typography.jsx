import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

const variantMapping = {
  h1: "h1",
  h2: "h2",
  h3: "h3",
  h4: "h4",
  h5: "h5",
  h6: "h6",
  subtitle1: "h6",
  subtitle2: "h6",
  body1: "p",
  body2: "p"
};

class Typography extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf([
      "initial",
      "inherit",
      "primary",
      "secondary",
      "tertiary",
      "positive",
      "negative"
    ]),

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     * By default, it maps the variant to a good default headline component.
     */
    component: PropTypes.elementType,

    /**
     * Controls the display type.
     */
    display: PropTypes.oneOf(["initial", "block", "inline"]),

    /**
     * If true, the text will have a bottom margin.
     */
    gutter: PropTypes.bool,

    /**
     * Set the text-align on the component.
     */
    textAlign: PropTypes.oneOf([
      "inherit",
      "left",
      "center",
      "right",
      "justify"
    ]),

    /**
     * Applies the theme typography styles.
     */
    variant: PropTypes.oneOf([
      "h1",
      "h2",
      "h3",
      "h4",
      "h5",
      "h6",
      "subtitle1",
      "subtitle2",
      "body1",
      "body2",
      "caption",
      "button",
      "inherit"
    ])
  };

  static defaultProps = {
    classes: [],
    color: "inherit",
    display: "initial",
    gutter: false,
    textAlign: "inherit",
    variant: "body1"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      color,
      component,
      display,
      gutter,
      textAlign,
      variant
    } = this.props;

    const theme = this.context;

    const className = clsx(
      "root",
      `color-${color}`,
      display,
      gutter ? "gutter" : null,
      variant,
      `align-${textAlign}`,
      classes
    );

    const Component = component || variantMapping[variant] || "span";

    return (
      <React.Fragment>
        <Component className={className}>{children}</Component>
        <style jsx>{`
          .root {
            margin: 0;
          }

          .block {
            display: block;
          }
          .inline {
            display: inline;
          }

          .gutter {
            margin-bottom: 0.35em;
          }

          .color-inherit {
            color: inherit;
          }
          .color-initial {
            color: initial;
          }
          .color-primary {
            color: ${theme.palette.primary.main};
          }
          .color-secondary {
            color: ${theme.palette.secondary.main};
          }
          .color-tertiary {
            color: ${theme.palette.tertiary.main};
          }
          .color-positive {
            color: ${theme.palette.positive.main};
          }
          .color-negative {
            color: ${theme.palette.negative.main};
          }

          .align-center {
            text-align: center;
          }
          .align-left {
            text-align: left;
          }
          .align-right {
            text-align: right;
          }

          .body1 {
            font-family: ${theme.typography.body1.fontFamily};
            font-weight: ${theme.typography.body1.fontWeight};
            font-size: ${theme.typography.body1.fontSize};
            line-height: ${theme.typography.body1.lineHeight};
            letter-spacing: ${theme.typography.body1.letterSpacing};
          }
          .body2 {
            font-family: ${theme.typography.body2.fontFamily};
            font-weight: ${theme.typography.body2.fontWeight};
            font-size: ${theme.typography.body2.fontSize};
            line-height: ${theme.typography.body2.lineHeight};
            letter-spacing: ${theme.typography.body2.letterSpacing};
          }
          .button {
            font-family: ${theme.typography.button.fontFamily};
            font-weight: ${theme.typography.button.fontWeight};
            font-size: ${theme.typography.button.fontSize};
            line-height: ${theme.typography.button.lineHeight};
            letter-spacing: ${theme.typography.button.letterSpacing};
          }
          .caption {
            font-family: ${theme.typography.caption.fontFamily};
            font-weight: ${theme.typography.caption.fontWeight};
            font-size: ${theme.typography.caption.fontSize};
            line-height: ${theme.typography.caption.lineHeight};
            letter-spacing: ${theme.typography.caption.letterSpacing};
          }
          .h1 {
            font-family: ${theme.typography.h1.fontFamily};
            font-weight: ${theme.typography.h1.fontWeight};
            font-size: ${theme.typography.h1.fontSize};
            line-height: ${theme.typography.h1.lineHeight};
            letter-spacing: ${theme.typography.h1.letterSpacing};
          }
          .h2 {
            font-family: ${theme.typography.h2.fontFamily};
            font-weight: ${theme.typography.h2.fontWeight};
            font-size: ${theme.typography.h2.fontSize};
            line-height: ${theme.typography.h2.lineHeight};
            letter-spacing: ${theme.typography.h2.letterSpacing};
          }
          .h3 {
            font-family: ${theme.typography.h3.fontFamily};
            font-weight: ${theme.typography.h3.fontWeight};
            font-size: ${theme.typography.h3.fontSize};
            line-height: ${theme.typography.h3.lineHeight};
            letter-spacing: ${theme.typography.h3.letterSpacing};
          }
          .h4 {
            font-family: ${theme.typography.h4.fontFamily};
            font-weight: ${theme.typography.h4.fontWeight};
            font-size: ${theme.typography.h4.fontSize};
            line-height: ${theme.typography.h4.lineHeight};
            letter-spacing: ${theme.typography.h4.letterSpacing};
          }
          .h5 {
            font-family: ${theme.typography.h5.fontFamily};
            font-weight: ${theme.typography.h5.fontWeight};
            font-size: ${theme.typography.h5.fontSize};
            line-height: ${theme.typography.h5.lineHeight};
            letter-spacing: ${theme.typography.h5.letterSpacing};
          }
          .subtitle1 {
            font-family: ${theme.typography.subtitle1.fontFamily};
            font-weight: ${theme.typography.subtitle1.fontWeight};
            font-size: ${theme.typography.subtitle1.fontSize};
            line-height: ${theme.typography.subtitle1.lineHeight};
            letter-spacing: ${theme.typography.subtitle1.letterSpacing};
          }
          .subtitle2 {
            font-family: ${theme.typography.subtitle2.fontFamily};
            font-weight: ${theme.typography.subtitle2.fontWeight};
            font-size: ${theme.typography.subtitle2.fontSize};
            line-height: ${theme.typography.subtitle2.lineHeight};
            letter-spacing: ${theme.typography.subtitle2.letterSpacing};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Typography);
