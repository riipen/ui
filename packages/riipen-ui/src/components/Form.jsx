import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Notice from "./Notice";
import Typography from "./Typography";

class Form extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * Top level form error to display.
     * An object consisting of key value errors or a single string
     * to display as an error.
     */
    error: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),

    /**
     * If `false`, pressing enter to submit the form will be disabled.
     */
    enter: PropTypes.bool
  };

  static defaultProps = {
    classes: [],
    enter: true
  };

  componentDidUpdate(prevProps) {
    const theme = this.context;

    // If an error id dynamically added to the form, scroll the error notice in to view
    if (!prevProps.error && this.props.error && this.error) {
      window.scrollTo(0, this.error.offsetTop - theme.spacing(2));
    }
  }

  setError = ref => {
    this.error = ref;
  };

  static contextType = ThemeContext;

  handleKeyPress = e => {
    const { enter } = this.props;

    if (e.key === "Enter" && !enter) {
      e.stopPropagation();
      e.preventDefault();
    }
  };

  renderErrorListItem = (key, message) => (
    <li key={key}>
      <Typography>
        {key}
        {": "}
        {message}
      </Typography>
    </li>
  );

  renderErrorList(error, key) {
    if (!error) return [];

    let newError;

    if (Array.isArray(error)) {
      newError = error.map((e, i) => {
        const newKey = key ? `${key} ${i}` : i;

        return this.renderErrorList(e, newKey);
      });
    } else if (typeof error === "object") {
      newError = Object.entries(error).map(([k, e]) => {
        const newKey = key ? `${key} ${k}` : k;

        return this.renderErrorList(e, newKey);
      });
    } else {
      newError = this.renderErrorListItem(key, error);
    }

    return newError;
  }

  renderErrorObject = error => <ul>{this.renderErrorList(error)}</ul>;

  renderErrorString = error => <Typography>{error}</Typography>;

  render() {
    const { children, classes, error, ...other } = this.props;

    const theme = this.context;

    const className = clsx(classes);

    // Remove any unwanted props from "other"
    other.enter = undefined;

    return (
      <React.Fragment>
        {error && (
          <div className={clsx("error")} ref={this.setError}>
            <Notice color="negative">
              {typeof error === "string" && this.renderErrorString(error)}
              {typeof error === "object" &&
                Object.keys(error).length > 0 &&
                this.renderErrorObject(error)}
            </Notice>
          </div>
        )}
        <form className={className} onKeyPress={this.handleKeyPress} {...other}>
          {children}
        </form>
        <style jsx>{`
          form {
            width: 100%;
          }

          .error {
            margin-bottom: ${theme.spacing(5)}px;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Form);
