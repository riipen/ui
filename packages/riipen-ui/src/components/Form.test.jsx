import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Form from "./Form";
import Typography from "./Typography";

describe("<Form>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Form />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Form />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Form />);

      expect(wrapper.find("Form").props().enter).toEqual(true);
      expect(wrapper.find("Form").props().errorScroll).toEqual(true);
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(<Form>{child}</Form>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<Form classes={classes} />);

      expect(
        wrapper
          .find("Form")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-form"];

      const wrapper = mount(<Form />);

      expect(wrapper.find("Form").props().classes).toEqual(classes.sort());
    });
  });

  describe("error prop", () => {
    const props = {};

    const tests = [
      { argument: "Error" },
      { argument: <Typography>Please provide a name</Typography> }
    ];

    tests.forEach(test => {
      it("sets class name as error with valid custom error type", () => {
        props.error = test.argument;

        const wrapper = mount(<Form error={props.error} />);

        expect(
          wrapper
            .find("Notice")
            .childAt(0)
            .childAt(0)
            .hasClass("error")
        ).toEqual(true);
      });
    });

    it("sets class name as errorContainer when valid error is used", () => {
      const error = "Yay";

      const wrapper = mount(<Form error={error} />);

      expect(
        wrapper
          .find("Form")
          .childAt(0)
          .hasClass("errorContainer")
      ).toEqual(true);
    });
  });

  describe("errors prop", () => {
    const props = {};

    const tests = [
      { argument: ["Error", "Sad"] },
      { argument: { errorOne: "Error", errorTwo: "Sad" } },
      {
        argument: [
          <Typography>Please provide a name</Typography>,
          <Typography>Error</Typography>
        ]
      }
    ];

    tests.forEach(test => {
      it("sets class name as errors with valid custom errors type", () => {
        props.errors = test.argument;

        const wrapper = mount(<Form errors={props.errors} />);

        expect(
          wrapper
            .find("Notice")
            .childAt(0)
            .childAt(0)
            .hasClass("errors")
        ).toEqual(true);
      });
    });

    it("sets class name as errorContainer when valid errors are used", () => {
      const errors = ["Error", "Sad"];

      const wrapper = mount(<Form errors={errors} />);

      expect(
        wrapper
          .find("Form")
          .childAt(0)
          .hasClass("errorContainer")
      ).toEqual(true);
    });

    it("does not set errorContainer class name without valid errors", () => {
      const wrapper = mount(<Form />);

      expect(
        wrapper
          .find("Form")
          .childAt(0)
          .hasClass("errorContainer")
      ).toEqual(false);
    });
  });
});
