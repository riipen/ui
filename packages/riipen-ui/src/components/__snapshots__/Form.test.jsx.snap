// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form> renders correct snapshot 1`] = `
<withClasses(Form)>
  <Form
    classes={
      Array [
        "riipen",
        "riipen-form",
      ]
    }
    enter={true}
    errorScroll={true}
  >
    <form
      className="jsx-846642530 riipen riipen-form"
      onKeyPress={[Function]}
    />
    <JSXStyle
      id="846642530"
    />
    <JSXStyle
      dynamic={
        Array [
          25,
          15,
          10,
        ]
      }
      id="3642758217"
    />
  </Form>
</withClasses(Form)>
`;
